src/txn/txn.c

```diff
@@ -1016,10 +1019,27 @@ __txn_fixup_prepared_update(
                     WT_STAT_CONN_INCR(
                       session, txn_prepare_rollback_fix_hs_update_with_ckpt_reserved_txnid);
                 }
+
+                /* The update can be inserted to the history store again. */
+                F_CLR(fix_upd, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
             } else
                 WT_STAT_CONN_INCR(session, txn_prepare_rollback_do_not_remove_hs_update);
-        } else
+        } else {
             WT_ERR(hs_cursor->remove(hs_cursor));
+            /* The update can be inserted to the history store again. */
```

Comment:
@agorrod @bvpvamsikrishna @kommiharibabu I think here we have another race. If we clear the flags after we remove the value from the history store and if we have a checkpoint starting and checkpoints this key in between, checkpoint will ignore the update restored because it still has the flags set, which is wrong.

If we clear the flags and then remove it from the history store, still a checkpoint can start to run and checkpoint the key in between, this time it will reinsert the update restored to the history store which hasn't been deleted and triggering out of order fix for that record. However, this out of order fix should not cause it to panic as it is the checkpoint that triggers it. May be that's acceptable?

Reply:
Do we need to do this at all? If the cleanup always happens during reconciliation we don't need to worry about racing threads.

Reply:
This is now resolved by the new solution.